flocker/control/_persistence.py

```diff
@@ -4,37 +4,65 @@
 Persistence of cluster configuration.
 """
 
-from pickle import dumps, loads
+from pyrsistent import PRecord, PVector, PMap, PSet
+from json import dumps, loads, JSONEncoder
 
+from twisted.python.filepath import FilePath
 from twisted.application.service import Service
 from twisted.internet.defer import succeed
 
-from ._model import Deployment
+from ._model import SERIALIZABLE_CLASSES, Deployment
 
 
-# These should not use Pickle!@!
-# https://clusterhq.atlassian.net/browse/FLOC-1241
-def serialize_deployment(deployment):
+class _ConfigurationEncoder(JSONEncoder):
     """
-    Convert a ``Deployment`` object to ``bytes``.
-
-    :param Deployment deployment: Object to serialize.
+    JSON encoder that can encode the configuration model.
+    """
+    def default(self, obj):
+        if isinstance(obj, PRecord):
+            result = dict(obj)
+            result[u"$__class__$"] = obj.__class__.__name__
+            return result
+        elif isinstance(obj, PMap):
+            return dict(obj)
+        elif isinstance(obj, (PSet, PVector, set)):
+            return list(obj)
+        elif isinstance(obj, FilePath):
+            return {u"$__class__$": u"FilePath",
+                    u"path": obj.path.decode("utf-8")}
+        return JSONEncoder.default(self, obj)
+
+
+def wire_encode(obj):
+    """
+    Encode the given configuration object into bytes.
 
-    :return bytes: Serialized object.
+    :param obj: An object from the configuration model, e.g. ``Deployment``.
+    :return bytes: Encoded object.
     """
-    return dumps(deployment)
+    return dumps(obj, cls=_ConfigurationEncoder)
 
 
-def deserialize_deployment(data):
+def wire_decode(data):
     """
-    Create a ``Deployment`` object that was previously serialized to given
-    ``bytes``.
+    Decode the given configuration object from bytes.
 
-    :param bytes data: Output of ``serialize_deployment``.
-
-    :return Deployment: Deserialized object.
+    :param bytes data: Encoded object.
+    :param obj: An object from the configuration model, e.g. ``Deployment``.
     """
-    return loads(data)
+    classes = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+    def decode_object(dictionary):
+        class_name = dictionary.get("$__class__$", None)
+        if class_name == u"FilePath":
+            return FilePath(dictionary.get(u"path").encode("utf-8"))
+        elif class_name in classes:
+            dictionary = dictionary.copy()
+            dictionary.pop("$__class__$")
+            return classes[class_name].create(dictionary)
+        else:
+            return dictionary
```

Comment:
Is there a test for disallowed classes not being deserialized?  I didn't see one.  There probably ought to be.